### ISSUE TITLE
Generic type of setter is leaking across instances

### ISSUE THREAD

**Issue by belm0** [PyCQA/pylint#8022]
### Bug description

Despite typing, the `value` setter below seems to remember the last assigned type, and applies it to an unrelated instance with different generic type.

```python
from typing import TypeVar, Generic

T = TypeVar('T')


class WrappedVal(Generic[T]):
    def __init__(self, value: T):
        self._value = value

    @property
    def value(self) -> T:
        return self._value

    @value.setter
    def value(self, x: T):
        self._value = x


val1 = WrappedVal('hello')
val1.value = 'goodbye'

val2 = WrappedVal(1.)
print(-val2.value)  # bad operand type for unary -: str
```

### Command used

```shell
pylint generic_setter.py
```

### Pylint output

```shell
generic_setter.py:23:6: E1130: bad operand type for unary -: str (invalid-unary-operand-type)
```

### Expected behavior

no error, since unary minus is valid for float

### Pylint version

```shell
pylint 2.15.9
astroid 2.12.13
Python 3.8.12 (default, Sep 23 2021, 08:42:37)
[Clang 12.0.0 (clang-1200.0.32.29)]
```

**Comment by belm0** [comment]
Defining the class without properties has the same problem, so it's not the setter.

```python
from typing import TypeVar, Generic

T = TypeVar('T')

class WrappedVal(Generic[T]):
    def __init__(self, value: T):
        self.value = value

val1 = WrappedVal('hello')
val1.value = 'goodbye'

val2 = WrappedVal(1.)
print(-val2.value)```

**Comment by mbyrnepr2** [comment]
Thanks @belm0. Can simplify the reproducing-code further with:
```python
class WrappedVal:
    def __init__(self, value):
        self.value = value

val1 = WrappedVal('hello')
val1.value = 'goodbye'

val2 = WrappedVal(1.)
print(-val2.value)  # bad operand type for unary -: str
```

**Comment by belm0** [comment]
Well, I think that error would be expected if generics aren't used, as the type inference is doing its best.

So if pylint behaves the same regardless of Generic typing, it means that
